[harvester doc] add doc of audit event (#208)
harvester/docs: docs/logging/logging.md

---
sidebar_position: 1
sidebar_label: Logging Audit Event
title: ""
keywords:
- Harvester
- Logging
- Audit
- Event
---

# Harvester Integration with Logging, Auditing and Event

_Available as of v1.1.0_

It is important to know what is happending / has happend in the `Harvester Cluster`.

`Harvester` collects the `cluster running log`, kubernetes `audit` and `event` log right after the cluster is powered on, which is helpful for monitoring, logging, auditing and troubleshooting.

`Harvester` supports sending those logs to various types of log servers.

:::note

> The size of logging data is related to the cluster scale, workload and other factors. `Harvester` does not use persistant storage to store log data inside the cluster. Users need to set up a log server to receive logs accordingly.

Reviewer: ```suggestion
The size of logging data is related to the cluster scale, workload and other factors. `Harvester` does not use persistent storage to store log data inside the cluster. Users need to set up a log server to receive logs accordingly.
```

---
sidebar_position: 1
sidebar_label: Logging Audit Event
title: ""
keywords:
- Harvester
- Logging
- Audit
- Event
---

> # Harvester Logging Audit Event

Reviewer 2: ```suggestion
# Harvester Integration with Logging, Auditing, and Event
```

_Available as of v1.1.0_

It is important to know what is happending / has happend in the `Harvester Cluster`.

`Harvester` collects the `cluster running log`, kubernetes `audit` and `event` log right after the cluster is powered on, which is helpful for monitoring, logging, auditing and troubleshooting.

> `Harvester` supports to send those logs to vairous types of log server.

Reviewer 3: ```suggestion
`Harvester` supports to send those logs to various types of log server.
```

typo nit

> `Harvester` supports to send those logs to vairous types of log server.

Reviewer: ```suggestion
`Harvester` supports sending those logs to various types of log servers.
```

:::note

> The logging related data size is related on the cluster scale, workload and other factors. `Harvester` does not use persistant storage to store log data inside the cluster. Users need to set up log server to receive logs accordingly.

Reviewer: ```suggestion
The logging related data size is related on the cluster scale, workload and other factors. `Harvester` does not use persistant storage to store log data inside the cluster. Users need to set up a log server to receive logs accordingly.
```

Reviewer 2: consider refactoring to:

The size of data associated with logs is related to cluster size, workload, and other factors. `Harvester` does not use persistent storage to store log data within the cluster. Users need to set up a log server to receive the corresponding logs.

:::

> ## General architecture

Reviewer 2: ```suggestion
## High-level Architecture
```

> `Harvester` integrates the `Banzai Cloud` `Logging Processing Pipeline`, adapts it with `Rancher` and `Harvester`.

Reviewer 2: ```suggestion
The [Banzai Cloud Logging operator](https://banzaicloud.com/docs/one-eye/logging-operator/) now powers both Harvester and Rancher as an in-house logging solution.
```

![](/img/v1.1/logging/fluent-operator.png)

> In Harvester's practice, the `Logging`, `Audit` and `Event` are sharing one architecture, the `Logging` is the infrastructure, the `Audit` and `Event` are on top of it.

Reviewer 2: I think this is something related to the HEP, for normal users, they probably don't need to understand the design here.

> In Harvester's practice, the `Logging`, `Audit` and `Event` are sharing one architecture, the `Logging` is the infrastructure, the `Audit` and `Event` are on top of it.

Author: Remove the second, internal architecture diagram.

    url: http://loki-stack.cattle-logging-system.svc:3100
    extra_labels:
      logOutput: harvester-loki
```

## Audit

> Harvester collects Kubernetes `audit` and is able to send the `audit` to vairous types of log server.

Reviewer: ```suggestion
Harvester collects Kubernetes `audit` and is able to send the `audit` to various types of log servers.
```

> The policy file to guide `kube-apiserver` is [here] (https://github.com/harvester/harvester-installer/blob/master/pkg/config/templates/rke2-92-harvester-kube-audit-policy.yaml).

Reviewer: ```suggestion
The policy file to guide `kube-apiserver` is [here](https://github.com/harvester/harvester-installer/blob/master/pkg/config/templates/rke2-92-harvester-kube-audit-policy.yaml).
```

Reviewer 4: Suggest using permalink in case it's removed.

Author: Updated with permalink. thanks.

RequestResponse - log event metadata, request and response bodies. This does not apply for non-resource requests.
```

### Audit Log Format

#### Audit Log Format in Kubernetes

Kubernetes apiserver logs audit with following JSON format into a local file.

```
{
"kind":"Event",
"apiVersion":"audit.k8s.io/v1",
"level":"Metadata",
"auditID":"13d0bf83-7249-417b-b386-d7fc7c024583",
"stage":"RequestReceived",
"requestURI":"/apis/flowcontrol.apiserver.k8s.io/v1beta2/prioritylevelconfigurations?fieldManager=api-priority-and-fairness-config-producer-v1",
"verb":"create",
"user":{"username":"system:apiserver","uid":"d311c1fe-2d96-4e54-a01b-5203936e1046","groups":["system:masters"]},
"sourceIPs":["::1"],
"userAgent":"kube-apiserver/v1.24.7+rke2r1 (linux/amd64) kubernetes/e6f3597",
"objectRef":{"resource":"prioritylevelconfigurations",
"apiGroup":"flowcontrol.apiserver.k8s.io",
"apiVersion":"v1beta2"},
"requestReceivedTimestamp":"2022-10-19T18:55:07.244781Z",
"stageTimestamp":"2022-10-19T18:55:07.244781Z"
}
```

#### Audit Log Format before Being Sent to Log Servers

Harvester keeps the `audit` log unchanged before sending it to the log server.

> ### Audit Log Output/Clusteroutput

Reviewer 4: ```suggestion
### Audit Log Output/ClusterOutput
```
Do we need the upper case for consistency?

Author: Good catch, updated with unified format. thanks.

The policy file to guide `kube-apiserver`:

https://github.com/harvester/harvester-installer/blob/master/pkg/config/templates/rke2-92-harvester-kube-audit-policy.yaml

### Audit definition

In `kubernetes`, the `audit` data is generated by `kube-apiserver` according to defined policy.

https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/

```
...
Audit policy
Audit policy defines rules about what events should be recorded and what data they should include. The audit policy object structure is defined in the audit.k8s.io API group. When an event is processed, it's compared against the list of rules in order. The first matching rule sets the audit level of the event. The defined audit levels are:

None - don't log events that match this rule.
Metadata - log request metadata (requesting user, timestamp, resource, verb, etc.) but not request or response body.
Request - log event metadata and request body but not response body. This does not apply for non-resource requests.
RequestResponse - log event metadata, request and response bodies. This does not apply for non-resource requests.
```

### Audit log format

#### Audit log format in kubernetes

> Kubernetes apiserver logs audit with following JSON format into a local file.

Author: It is kept. 

There is no `sending` action, the `kube apiserver` write to a file directly by itself.

> > {"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"13d0bf83-7249-417b-b386-d7fc7c024583","stage":"RequestReceived","requestURI":"/apis/flowcontrol.apiserver.k8s.io/v1beta2/prioritylevelconfigurations?fieldManager=api-priority-and-fairness-config-producer-v1","verb":"create","user":{"username":"system:apiserver","uid":"d311c1fe-2d96-4e54-a01b-5203936e1046","groups":["system:masters"]},"sourceIPs":["::1"],"userAgent":"kube-apiserver/v1.24.7+rke2r1 (linux/amd64) kubernetes/e6f3597","objectRef":{"resource":"prioritylevelconfigurations","apiGroup":"flowcontrol.apiserver.k8s.io","apiVersion":"v1beta2"},"requestReceivedTimestamp":"2022-10-19T18:55:07.244781Z","stageTimestamp":"2022-10-19T18:55:07.244781Z"}

Reviewer 3: Should this be "pretty printed" so its more human-readable?

#### Audit log format before sending to log server

Harvester keeps the `audit` log unchanged before sending to log server.

> {"kind":"Event","apiVersion":"audit.k8s.io/v1","level":"Metadata","auditID":"13d0bf83-7249-417b-b386-d7fc7c024583","stage":"RequestReceived","requestURI":"/apis/flowcontrol.apiserver.k8s.io/v1beta2/prioritylevelconfigurations?fieldManager=api-priority-and-fairness-config-producer-v1","verb":"create","user":{"username":"system:apiserver","uid":"d311c1fe-2d96-4e54-a01b-5203936e1046","groups":["system:masters"]},"sourceIPs":["::1"],"userAgent":"kube-apiserver/v1.24.7+rke2r1 (linux/amd64) kubernetes/e6f3597","objectRef":{"resource":"prioritylevelconfigurations","apiGroup":"flowcontrol.apiserver.k8s.io","apiVersion":"v1beta2"},"requestReceivedTimestamp":"2022-10-19T18:55:07.244781Z","stageTimestamp":"2022-10-19T18:55:07.244781Z"}

### Audit log output/clusteroutput

> To output audit related log, the `output`/`clusteroutput` needs a `loggingRef` field with fixed value `harvester-kube-audit-log-ref`.

Reviewer: ```suggestion
To output audit related log, the `output`/`clusteroutput` requires the value of `loggingRef` to be `harvester-kube-audit-log-ref`.
```

> When users config from Harvester dashboard, the field is added automatically.
> 
> Select type `Audit Only`.

Reviewer: ```suggestion
When you configure from the Harvester dashboard, the field is added automatically.

Select `Audit Only` from the `Type` drop-down list.
```

![](/img/v1.1/logging/cluster-output-type.png)

> When users config from CLI, please add the field accordingly.

Reviewer: ```suggestion
When you configure from the CLI, please add the field manually.
```

Example:

```
apiVersion: logging.banzaicloud.io/v1beta1
kind: ClusterOutput
metadata:
  name: "harvester-audit-webhook"
  namespace: "cattle-logging-system"
spec:
  http:
    endpoint: "http://192.168.122.159:8096/"
    open_timeout: 3
    format: 
      type: "json"
    buffer:
      chunk_limit_size: 3MB
      timekey: 2m
      timekey_wait: 1m
  loggingRef: harvester-kube-audit-log-ref   # this reference is fixed and must be here
```

### Audit log flow/clusterflow

> To route audit related log, the `flow`/`clusterflow` needs a `loggingRef` field with fixed value `harvester-kube-audit-log-ref`.

Reviewer: ```suggestion
To route audit related logs, the `flow`/`clusterflow` requires the value of `loggingRef` to be `harvester-kube-audit-log-ref`.
```

> When users config from Harvester dashboard, the field is added automatically.

Reviewer: ```suggestion
When you configure from the Harvester dashboard, the field is added automatically.
```

Select type `Audit`.

![](/img/v1.1/logging/cluster-flow-type.png)

> When users config from CLI, please add the field accordingly.

Reviewer: ```suggestion
When you configure from the CLI, please add the field manually.
```

Example:

```
apiVersion: logging.banzaicloud.io/v1beta1
kind: ClusterFlow
metadata:
  name: "harvester-audit-webhook"
  namespace: "cattle-logging-system"
spec:
  globalOutputRefs:
    - "harvester-audit-webhook"
  loggingRef: harvester-kube-audit-log-ref  # this reference is fixed and must be here
```

### Harvester

## Event

> Harvester collects Kubernetes `event` and is able to send the `event` to vairous types of log server.

Reviewer: ```suggestion
Harvester collects Kubernetes `event` and is able to send the `event` to various types of log servers.
```

### Event definition

Kubernetes `events` are objects that show you what is happening inside a cluster, such as what decisions were made by the scheduler or why some pods were evicted from the node. All core components and extensions (operators/controllers) may create events through the API Server.

> Events have no direct relationship with log messages generated by the various components, and are not affected with the log verbosity level. When a component creates an event, it often emits a corresponding log message. Events are garbage collected by the API Server after a short time (typically after an hour, 5 minutes in case of EKS), which means that they can be used to understand issues just happening, but you have to collect them to investigate past events.

Reviewer: ```suggestion
Events have no direct relationship with log messages generated by the various components, and are not affected by the log verbosity level. When a component creates an event, it often emits a corresponding log message. Events are garbage collected by the API Server after a short time (typically after an hour, or 5 minutes in the case of EKS), which means that they can be used to understand issues that are happening, but you have to collect them to investigate past events.
```

```
        {
            "apiVersion": "v1",
            "count": 1,
            "eventTime": null,
            "firstTimestamp": "2022-08-24T11:17:35Z",
            "involvedObject": {
                "apiVersion": "kubevirt.io/v1",
                "kind": "VirtualMachineInstance",
                "name": "vm-ide-1",
                "namespace": "default",
                "resourceVersion": "604601",
                "uid": "1bd4133f-5aa3-4eda-bd26-3193b255b480"
            },
            "kind": "Event",
            "lastTimestamp": "2022-08-24T11:17:35Z",
            "message": "VirtualMachineInstance defined.",
            "metadata": {
                "creationTimestamp": "2022-08-24T11:17:35Z",
                "name": "vm-ide-1.170e43cbdd833b62",
                "namespace": "default",
                "resourceVersion": "604626",
                "uid": "0114f4e7-1d4a-4201-b0e5-8cc8ede202f4"
            },
            "reason": "Created",
            "reportingComponent": "",
            "reportingInstance": "",
            "source": {
                "component": "virt-handler",
                "host": "harv1"
            },
            "type": "Normal"
        },
```

> #### Event log format before sending to log server

Reviewer: ```suggestion
#### Event log format before being sent to log servers
```

> Each `event log` has format of: `{"stream":"","logtag":"F","message":"","kubernetes":{""}}`. The `kubernetes event` is in the field `message`.

Reviewer: ```suggestion
Each `event log` has the format of: `{"stream":"","logtag":"F","message":"","kubernetes":{""}}`. The `kubernetes event` is in the field `message`.
```

}

```

### Event log output/clusteroutput

> Event share the `output`/`clusteroutput` with `Logging`.

Reviewer: ```suggestion
Events share the `output`/`clusteroutput` with `Logging`.
```

> Select type `Logging/Event`.

Reviewer: ```suggestion
Select `Logging/Event` from the `Type` drop-down list.
```

![](/img/v1.1/logging/cluster-output-type.png)

### Event log flow/clusterflow

> Comparing to normal Logging `flow`/`clusterflow`, the `Event` related `flow`/`clusterflow`, has a more match field with value `event-tailer`.

Reviewer: ```suggestion
Compared with the normal Logging `flow`/`clusterflow`, the `Event` related `flow`/`clusterflow` has one more match field with the value of `event-tailer`.
```

Reviewer: for `has one more match field`, should we also point out which match field this is?

> When users config from Harvester dashboard, the field is added automatically.
> 
> Select type `Event`.
> 
> ![](/img/v1.1/logging/cluster-flow-type.png)
> 
> When users config from CLI, please add the field accordingly.

Reviewer: ```suggestion
When you configure from the Harvester dashboard, the field is added automatically.

Select `Event` from the `Type` drop-down list.

![](/img/v1.1/logging/cluster-flow-type.png)

When you configure from the CLI, please add the field manually.
```
